debian/bongorestore.1
.\" Man page generated from reStructuredText.
.
.TH "BONGORESTORE" "1" "January 30, 2015" "3.0" "bongodb-manual"
.SH NAME
bongorestore \- BongoDB Data Restoration Tool
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
The \fBbongorestore\fP program writes data from a binary database
dump created by \fBbongodump\fP to a BongoDB
instance. \fBbongorestore\fP can create a new database or add
data to an existing database.
.sp
\fBbongorestore\fP can write data to either \fIbongod\fP or \fBbongos\fP
instances, in addition to writing directly to BongoDB data files
without an active \fBbongod\fP\&.
.SH BEHAVIOR
.sp
If you restore to an existing database, \fBbongorestore\fP will
only insert into the existing database, and does not perform updates
of any kind. If existing documents have the same value \fB_id\fP field
in the target database and collection,
\fBbongorestore\fP will \fInot\fP overwrite those documents.
.sp
Remember the following properties of \fBbongorestore\fP behavior:
.INDENT 0.0
.IP \(bu 2
\fBbongorestore\fP recreates indexes recorded by
\fBbongodump\fP\&.
.IP \(bu 2
all operations are inserts, not updates.
.IP \(bu 2
\fBbongorestore\fP does not wait for a response from a
\fBbongod\fP to ensure that the BongoDB process has received or
recorded the operation.
.sp
The \fBbongod\fP will record any errors to its log that occur
during a restore operation, but \fBbongorestore\fP will not
receive errors.
.UNINDENT
.sp
The data format used by \fBbongodump\fP from version 2.2 or
later is \fIincompatible\fP with earlier versions of \fBbongod\fP\&.
Do not use recent versions of \fBbongodump\fP to back up older
data stores.
.sp
New in version 3.0.0: \fBbongorestore\fP also accepts input via standard input.

.SH REQUIRED ACCESS TO RESTORE USER DATA
.sp
Changed in version 2.6.

.sp
To restore users and \fIuser\-defined roles\fP on a
given database, you must have access to the \fBadmin\fP database. BongoDB
stores the user data and role definitions for all databases in the
\fBadmin\fP database.
.sp
Specifically, to restore users to a given database, you must have the
\fBinsert\fP \fIaction\fP on the \fBadmin\fP
database\(aqs \fBadmin.system.users\fP collection. The \fBrestore\fP
role provides this privilege.
.sp
To restore user\-defined roles to a database, you must have the
\fBinsert\fP action on the \fBadmin\fP database\(aqs
\fBadmin.system.roles\fP collection. The \fBrestore\fP role
provides this privilege.
.SH OPTIONS
.INDENT 0.0
.TP
.B bongorestore
.UNINDENT
.INDENT 0.0
.TP
.B bongorestore
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBbongorestore\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Runs the \fBbongorestore\fP in a quiet mode that attempts to limit the amount
of output.
.sp
This option suppresses:
.INDENT 7.0
.IP \(bu 2
output from \fIdatabase commands\fP
.IP \(bu 2
replication activity
.IP \(bu 2
connection accepted events
.IP \(bu 2
connection closed events
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBbongorestore\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname><:port>, \-h <hostname><:port>
\fIDefault\fP: localhost:27017
.sp
Specifies a resolvable hostname for the \fBbongod\fP to which to
connect. By default, the \fBbongorestore\fP attempts to connect to a BongoDB
instance running on the localhost on port number \fB27017\fP\&.
.sp
To connect to a replica set, specify the
\fBreplSetName\fP and a seed list of set members, as in
the following:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
<replSetName>/<hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can always connect directly to a single BongoDB instance by
specifying the host and port number directly.
.sp
Changed in version 3.0.0: If you use IPv6 and use the \fB<address>:<port>\fP format, you must
enclose the portion of an address and port combination in
brackets (e.g. \fB[<address>]\fP).

.UNINDENT
.INDENT 0.0
.TP
.B \-\-port <port>
\fIDefault\fP: 27017
.sp
Specifies the TCP port on which the BongoDB instance listens for
client connections.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
Enables IPv6 support and allows the \fBbongorestore\fP to connect to the
BongoDB instance using an IPv6 network. All BongoDB programs and
processes disable IPv6 support by default.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ssl
New in version 2.6.

.sp
Enables connection to a \fBbongod\fP or \fBbongos\fP that has
SSL support enabled.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCAFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the root certificate chain
from the Certificate Authority. Specify the file name of the
\fB\&.pem\fP file using relative or absolute paths.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
If the \fBbongo\fP shell or any other tool that connects to
\fBbongos\fP or \fBbongod\fP is run without
\fI\-\-sslCAFile\fP, it will not attempt to validate
server certificates. This results in vulnerability to expired
\fBbongod\fP and \fBbongos\fP certificates as well as to foreign
processes posing as valid \fBbongod\fP or \fBbongos\fP
instances. Ensure that you \fIalways\fP specify the CA file against which
server certificates should be validated in cases where intrusion is a
possibility.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains both the SSL certificate
and key. Specify the file name of the \fB\&.pem\fP file using relative
or absolute paths.
.sp
This option is required when using the \fI\-\-ssl\fP option to connect
to a \fBbongod\fP or \fBbongos\fP that has
\fBCAFile\fP enabled \fIwithout\fP
\fBallowConnectionsWithoutCertificates\fP\&.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyPassword <value>
New in version 2.6.

.sp
Specifies the password to de\-crypt the certificate\-key file (i.e.
\fI\-\-sslPEMKeyFile\fP). Use the \fI\-\-sslPEMKeyPassword\fP option only if the
certificate\-key file is encrypted. In all cases, the \fBbongorestore\fP will
redact the password from all logging and reporting output.
.sp
If the private key in the PEM file is encrypted and you do not specify
the \fI\-\-sslPEMKeyPassword\fP option, the \fBbongorestore\fP will prompt for a passphrase. See
\fIssl\-certificate\-password\fP\&.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCRLFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the Certificate Revocation
List. Specify the file name of the \fB\&.pem\fP file using relative or
absolute paths.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidCertificates
New in version 2.6.

.sp
Bypasses the validation checks for server certificates and allows
the use of invalid certificates. When using the
\fBallowInvalidCertificates\fP setting, BongoDB logs as a
warning the use of the invalid certificate.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidHostnames
New in version 3.0.

.sp
Disables the validation of the hostnames in SSL certificates. Allows
\fBbongorestore\fP to connect to BongoDB instances if the hostname their
certificates do not match the specified hostname.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslFIPSMode
New in version 2.6.

.sp
Directs the \fBbongorestore\fP to use the FIPS mode of the installed OpenSSL
library. Your system must have a FIPS compliant OpenSSL library to use
the \fI\-\-sslFIPSMode\fP option.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
FIPS Compatible SSL is
available only in \fI\%BongoDB Enterprise\fP\&. See
http://docs.bongodb.org/manual/tutorial/configure\-fips for more information.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a BongoDB database
that uses authentication. Use in conjunction with the \fB\-\-password\fP and
\fB\-\-authenticationDatabase\fP options.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a BongoDB database
that uses authentication. Use in conjunction with the \fB\-\-username\fP and
\fB\-\-authenticationDatabase\fP options.
.sp
If you do not specify an argument for \fI\-\-password\fP, \fBbongorestore\fP will
prompt interactively for a password on the console.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationDatabase <dbname>
New in version 2.4.

.sp
Specifies the database that holds the user\(aqs credentials.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationMechanism <name>
\fIDefault\fP: BONGODB\-CR
.sp
New in version 2.4.

.sp
Changed in version 2.6: Added support for the \fBPLAIN\fP and \fBBONGODB\-X509\fP authentication
mechanisms.

.sp
Specifies the authentication mechanism the \fBbongorestore\fP instance uses to
authenticate to the \fBbongod\fP or \fBbongos\fP\&.
.TS
center;
|l|l|.
_
T{
Value
T}	T{
Description
T}
_
T{
BONGODB\-CR
T}	T{
BongoDB challenge/response authentication.
T}
_
T{
BONGODB\-X509
T}	T{
BongoDB SSL certificate authentication.
T}
_
T{
PLAIN
T}	T{
External authentication using LDAP. You can also use \fBPLAIN\fP
for authenticating in\-database users. \fBPLAIN\fP transmits
passwords in plain text. This mechanism is available only in
\fI\%BongoDB Enterprise\fP\&.
T}
_
T{
GSSAPI
T}	T{
External authentication using Kerberos. This mechanism is
available only in \fI\%BongoDB Enterprise\fP\&.
T}
_
.TE
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiServiceName
New in version 2.6.

.sp
Specify the name of the service using \fBGSSAPI/Kerberos\fP\&. Only required if the service does not use the
default name of \fBbongodb\fP\&.
.sp
This option is available only in BongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiHostName
New in version 2.6.

.sp
Specify the hostname of a service using \fBGSSAPI/Kerberos\fP\&. \fIOnly\fP required if the hostname of a machine does
not match the hostname resolved by DNS.
.sp
This option is available only in BongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-db <database>, \-d <database>
Specifies a database for \fBbongorestore\fP to restore data \fIinto\fP\&.
If the database does not exist, \fBbongorestore\fP creates the
database. If you do not specify a \fB<db>\fP, \fBbongorestore\fP
creates new databases that correspond to the databases where data
originated and data may be overwritten. Use this option to restore data
into a BongoDB instance that already has data.
.sp
\fI\%\-\-db\fP does \fInot\fP control which \fIBSON\fP files
\fBbongorestore\fP restores. You must use the
\fBbongorestore\fP \fI\%path option\fP to
limit that restored data.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-collection <collection>, \-c <collection>
Specifies a single collection for \fBbongorestore\fP to restore. If
you do not specify \fI\%\-\-collection\fP, \fBbongorestore\fP takes
the collection name from the input filename. If the input file has an
extension, BongoDB omits the extension of the file from the collection
name.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-objcheck
Forces \fBbongorestore\fP to validate all requests from clients
upon receipt to ensure that clients never insert invalid documents into
the database. For objects with a high degree of sub\-document nesting,
\fI\-\-objcheck\fP can have a small impact on performance. You can set
\fI\%\-\-noobjcheck\fP to disable object checking at run\-time.
.sp
Changed in version 2.4: BongoDB enables \fI\-\-objcheck\fP by default, to prevent any
client from inserting malformed or invalid BSON into a BongoDB
database.

.UNINDENT
.INDENT 0.0
.TP
.B \-\-noobjcheck
New in version 2.4.

.sp
Disables the default document validation that BongoDB performs on all
incoming BSON documents.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-drop
Modifies the restoration procedure to drop every collection from the
target database before restoring the collection from the dumped backup.
.sp
With \fI\%\-\-drop\fP specified, \fBbongorestore\fP removes all user
credentials and replaces them with users defined in the dump
file. Therefore, in systems with \fBauthorization\fP
enabled, \fBbongorestore\fP must be able to authenticate to an existing
user \fIand\fP to a user defined in the dump file. If \fBbongorestore\fP can\(aqt
authenticate to a user defined in the dump file, the restoration
process will fail, leaving an empty database.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-oplogReplay
Replays the \fIoplog\fP after restoring the dump to ensure that the
current state of the database reflects the point\-in\-time backup captured
with the "\fIbongodump \-\-oplog\fP" command. For an example of
\fI\%\-\-oplogReplay\fP, see \fIbackup\-restore\-oplogreplay\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-oplogLimit <timestamp>
New in version 2.2.

.sp
Prevents \fBbongorestore\fP from applying \fIoplog\fP entries
with timestamp newer than or equal to \fB<timestamp>\fP\&. Specify
\fB<timestamp>\fP values in the form of \fB<time_t>:<ordinal>\fP, where
\fB<time_t>\fP is the seconds since the UNIX epoch, and \fB<ordinal>\fP
represents a counter of operations in the oplog that occurred in the
specified second.
.sp
You must use \fI\%\-\-oplogLimit\fP in conjunction with the
\fI\%\-\-oplogReplay\fP option.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-keepIndexVersion
Prevents \fBbongorestore\fP from upgrading the index to the latest
version during the restoration process.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-noIndexRestore
New in version 2.2.

.sp
Prevents \fBbongorestore\fP from restoring and building indexes as
specified in the corresponding \fBbongodump\fP output.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-noOptionsRestore
New in version 2.2.

.sp
Prevents \fBbongorestore\fP from setting the collection options,
such as those specified by the \fBcollMod\fP \fIdatabase
command\fP, on restored collections.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-restoreDbUsersAndRoles
Restore user and role definitions for the given database. See
http://docs.bongodb.org/manual/reference/system\-roles\-collection and
http://docs.bongodb.org/manual/reference/system\-users\-collection for more information.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-w <number of replicas per write>
New in version 2.2.

.sp
Specifies the \fIwrite concern\fP for each write operation that
\fBbongorestore\fP writes to the target database. By default,
\fBbongorestore\fP does not wait for a response for \fIwrite
acknowledgment\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-writeConcern <document>
\fIDefault\fP: majority
.sp
Specifies the \fIwrite concern\fP for each write operation that \fBbongorestore\fP
writes to the target database.
.sp
Specify the write concern as a document with \fIw options\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-maintainInsertionOrder
\fIDefault\fP: False
.sp
If specified, \fBbongorestore\fP inserts the documents in the order of
their appearance in the input source, otherwise \fBbongorestore\fP may
perform the insertions in an arbitrary order.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-numParallelCollections int, \-j int
\fIDefault\fP: 4
.sp
Number of collections \fBbongorestore\fP should restore
in parallel.
.UNINDENT
.INDENT 0.0
.TP
.B <path>
The final argument of the \fBbongorestore\fP command is a
directory path. This argument specifies the location of the
database dump from which to restore.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-dir string
Specifies the dump directory.
.UNINDENT
.SH USE
.sp
See http://docs.bongodb.org/manual/tutorial/backup\-with\-bongodump
for a larger overview of \fBbongorestore\fP
usage. Also see the \fBbongodump\fP document for an overview of the
\fBbongodump\fP, which provides the related inverse
functionality.
.sp
Consider the following example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongorestore \-\-collection people \-\-db accounts dump/accounts/people.bson
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Here, \fBbongorestore\fP reads the database dump in the \fBdump/\fP
sub\-directory of the current directory, and restores \fIonly\fP the
documents in the collection named \fBpeople\fP from the database named
\fBaccounts\fP\&. \fBbongorestore\fP restores data to the instance
running on the localhost interface on port \fB27017\fP\&.
.sp
In the final example, \fBbongorestore\fP restores a database
dump located at \fB/opt/backup/bongodump\-2011\-10\-24\fP, to a database
running on port \fB37017\fP on the host
\fBbongodb1.example.net\fP\&. The \fBbongorestore\fP command authenticates to
the BongoDB instance using the username \fBuser\fP and the
password \fBpass\fP, as follows:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongorestore \-\-host bongodb1.example.net \-\-port 37017 \-\-username user \-\-password pass /opt/backup/bongodump\-2011\-10\-24
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can also \fIpipe\fP data directly into to \fBbongorestore\fP
through standard input, as in the following example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
zcat /opt/backup/bongodump\-2014\-12\-03/accounts.people.bson.gz | bongorestore \-\-collection people \-\-db accounts
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
BongoDB Documentation Project
.SH COPYRIGHT
2011-2015
.\" Generated by docutils manpage writer.
.


NEWS
WiredTiger release 2.9.0, 2016-09-06
------------------------------------

New features and API changes; refer to the API documentation for full details:

* WT-2360	Add nesting and disjunction functionality to cursor joins
* WT-2552	Add a public API to allow custom filesystem implementations
* WT-2738	Remove the ability to change the default checkpoint name. See the upgrading documentation for details.
* WT-2711	Change the statistics log configuration options. Restrict where statistics log files can be created to remove security vulnerability. See upgrading documentation for details.
* WT-2764	Enhance the checkpoint implementation, which led to a change in default eviction settings. See the upgrading documentation for details.
* WT-2880	Add Zstandard compression support
* See the upgrading documentation for information about renamed and removed statistics.

Significant changes and bug fixes:

* WT-2026	Fix a bug where in-memory pages were allowed to grow too large
* WT-2343	Ensure we cannot rename/drop when a backup cursor is open
* WT-2520	Ensure WT_SESSION::verify does not alter tables
* WT-2576	Ensure variable-length column-store cannot out-of-order return
* WT-2664	Change eviction so any eviction thread can find candidates
* WT-2702	Resolve an issue where under high thread load, WiredTiger exceeds cache size
* WT-2731	Resolve an issue where raw compression can create pages that are larger than expected
* WT-2737	Allow reconciliation to scrub dirty pages rather than evicting them
* WT-2757	Fix a bug in column stores where it was not possible to retrieve the new record number after an append
* WT-2816	Improve eviction performance

Issues fixed in MongoDB:

* SERVER-23504	Resolve a resource leak
* SERVER-23588	Stop using _open_osfhandle on Windows
* SERVER-23659	Improve log error message at startup
* SERVER-23661	Fix an issue where $sample takes disproportionately long time on newly created collection
* SERVER-24306	Fix stall in log_flush switching to new files
* SERVER-24580	Add more eviction stats to track efficiency
* SERVER-24580	Enhance eviction when application threads are contributing
* SERVER-24971	Don't cache buffers after application eviction
* SERVER-25843	Remove a redundant test
* SERVER-25845	Resolve an explicit null dereferenced
* SERVER-25846	Resolve a dereference after null with a check
* SERVER-26753	Don't spin on a read-lock in a tight loop

Other noteworthy changes since the previous release:

* WT-2103	Add incremental backup testing to test/format
* WT-2330	Ensure In-memory configurations do not create on-disk collection files
* WT-2450	Resolve an issue where salvage releases pages, then explicitly evicts them. Which can race with the eviction server
* WT-2453	Add multiple eviction queues
* WT-2504	Remove READONLY conditional for base config
* WT-2505	Resolve clang analyzer warnings
* WT-2508	Ensure test programs should remove test directories on the "clean" target
* WT-2512	Change how wtperf throttle values per thread are calculated
* WT-2513	Fix a type conversion warning
* WT-2517	Resolve incorrect uses of setvbuf on Windows
* WT-2518	Optimization of LSM checkpoint handle acquisition
* WT-2522	Resolve incorrect format code in message
* WT-2525	Miscellaneous cleanups of In-memory configurations
* WT-2526	Prevent mixing and matching readonly file-handles and read/write data-handles
* WT-2528	Fix a style error in WiredTiger build
* WT-2529	Be less aggressive asserting in readonly connections
* WT-2531	Resolve a case where in-memory tables are wasting space in truncation
* WT-2532	Change the WT_STREAM_APPEND and WT_STREAM_LINE_BUFFER flags as they overlap
* WT-2533	Don't let in-memory tables return a zero size
* WT-2534	Use atomic add when allocating transaction IDs
* WT-2535	Extend test/format to test for transactions reading their writes
* WT-2537	Check for an old and new uninitialized LSN for recovery
* WT-2539	Make streams a separate handle type from files
* WT-2540	Separate stream and file handle methods
* WT-2542	Resolve fixed-length column store reconciliation overwriting original values
* WT-2544	Fix eviction stats when clear is used
* WT-2546	Change evict server to be smarter when it decides whether or not to evict pages
* WT-2547	Add eviction testing configurations for 1 thread
* WT-2548	Cap the amount of data handed to raw compression
* WT-2550	Add support for raw mode cursors in java
* WT-2553	Add some In-memory documentation and stress testing
* WT-2556	Fix a typo in the Java example code
* WT-2557	Ensure that we truncate logs before closing backup cursor
* WT-2558	Reorganize WT_PAGE structure
* WT-2559	Resolve a potential segfault in Windows logging code
* WT-2560	Add a rwlock to protect transaction state, don't spin on this lock
* WT-2562	Add sleep loop to adapt for slow machines like PPC
* WT-2565	Further limit the amount of data passed to raw compression
* WT-2566	Add explicit memory barriers to Lock/unlock operations
* WT-2567	Add a check to see if logging is enabled for truncate rather than assuming it is
* WT-2568	Add a backward compatible constructor for PackInputStream
* WT-2569	Ensure __win_handle_read is followed by a call to GetLastError()
* WT-2570	Minor lint and cleanups to code
* WT-2571	Join code clean up
* WT-2572	Ensure test/format doesn't select an incompatible set of options for in-memory options configured
* WT-2573	Resolve a potential case where a double free can occur
* WT-2574	Ensure test/format frees all allocated configure memory
* WT-2576	Ensure variable-length column-store cannot out-of-order return
* WT-2577	Resolve a potential double-free following an in-memory split
* WT-2579	Change so diagnostic only code can use file streams directly
* WT-2580	Fix potential SWIG naming conflict in Java
* WT-2581	Resolve a case where we fail to track saved updates during a page split
* WT-2582	Fix a case where eviction can return WT_RESTART when running in-memory
* WT-2584	Change to not use periods at the end of error messages

src/bongo/gotools/bongorestore/main/bongorestore.go
// Main package for the bongorestore tool.
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/bongodb/bongo-tools/common/db"
	"github.com/bongodb/bongo-tools/common/log"
	"github.com/bongodb/bongo-tools/common/options"
	"github.com/bongodb/bongo-tools/common/progress"
	"github.com/bongodb/bongo-tools/common/signals"
	"github.com/bongodb/bongo-tools/common/util"
	"github.com/bongodb/bongo-tools/bongorestore"
)

const (
	progressBarLength   = 24
	progressBarWaitTime = time.Second * 3
)

func main() {
	// initialize command-line opts
	opts := options.New("bongorestore", bongorestore.Usage,
		options.EnabledOptions{Auth: true, Connection: true})
	nsOpts := &bongorestore.NSOptions{}
	opts.AddOptions(nsOpts)
	inputOpts := &bongorestore.InputOptions{}
	opts.AddOptions(inputOpts)
	outputOpts := &bongorestore.OutputOptions{}
	opts.AddOptions(outputOpts)

	extraArgs, err := opts.Parse()
	if err != nil {
		log.Logvf(log.Always, "error parsing command line options: %v", err)
		log.Logvf(log.Always, "try 'bongorestore --help' for more information")
		os.Exit(util.ExitBadOptions)
	}

	// Allow the db connector to fall back onto the current database when no
	// auth database is given; the standard -d/-c options go into nsOpts now
	opts.Namespace = &options.Namespace{DB: nsOpts.DB}

	// print help or version info, if specified
	if opts.PrintHelp(false) {
		return
	}

	if opts.PrintVersion() {
		return
	}

	log.SetVerbosity(opts.Verbosity)

	targetDir, err := getTargetDirFromArgs(extraArgs, inputOpts.Directory)
	if err != nil {
		log.Logvf(log.Always, "%v", err)
		log.Logvf(log.Always, "try 'bongorestore --help' for more information")
		os.Exit(util.ExitBadOptions)
	}
	targetDir = util.ToUniversalPath(targetDir)

	// connect directly, unless a replica set name is explicitly specified
	_, setName := util.ParseConnectionString(opts.Host)
	opts.Direct = (setName == "")
	opts.ReplicaSetName = setName

	provider, err := db.NewSessionProvider(*opts)
	if err != nil {
		log.Logvf(log.Always, "error connecting to host: %v", err)
		os.Exit(util.ExitError)
	}
	defer provider.Close()
	provider.SetBypassDocumentValidation(outputOpts.BypassDocumentValidation)

	// disable TCP timeouts for restore jobs
	provider.SetFlags(db.DisableSocketTimeout)

	// start up the progress bar manager
	progressManager := progress.NewBarWriter(log.Writer(0), progressBarWaitTime, progressBarLength, true)
	progressManager.Start()
	defer progressManager.Stop()

	restore := bongorestore.BongoRestore{
		ToolOptions:     opts,
		OutputOptions:   outputOpts,
		InputOptions:    inputOpts,
		NSOptions:       nsOpts,
		TargetDirectory: targetDir,
		SessionProvider: provider,
		ProgressManager: progressManager,
	}

	finishedChan := signals.HandleWithInterrupt(restore.HandleInterrupt)
	defer close(finishedChan)

	if err = restore.Restore(); err != nil {
		log.Logvf(log.Always, "Failed: %v", err)
		if err == util.ErrTerminated {
			os.Exit(util.ExitKill)
		}
		os.Exit(util.ExitError)
	}
}

// getTargetDirFromArgs handles the logic and error cases of figuring out
// the target restore directory.
func getTargetDirFromArgs(extraArgs []string, dirFlag string) (string, error) {
	// This logic is in a switch statement so that the rules are understandable.
	// We start by handling error cases, and then handle the different ways the target
	// directory can be legally set.
	switch {
	case len(extraArgs) > 1:
		// error on cases when there are too many positional arguments
		return "", fmt.Errorf("too many positional arguments")

	case dirFlag != "" && len(extraArgs) > 0:
		// error when positional arguments and --dir are used
		return "", fmt.Errorf(
			"cannot use both --dir and a positional argument to set the target directory")

	case len(extraArgs) == 1:
		// a nice, simple case where one argument is given, so we use it
		return extraArgs[0], nil

	case dirFlag != "":
		// if we have no extra args and a --dir flag, use the --dir flag
		log.Logv(log.Info, "using --dir flag instead of arguments")
		return dirFlag, nil

	default:
		return "", nil
	}
}
